ompi/mca/op/aarch64/op_aarch64_functions.c

```diff
@@ -0,0 +1,641 @@
+        type##type_size##_t *in = (type##type_size##_t *) _in,                                \
+                            *out = (type##type_size##_t *) _out;                              \
+        OP_CONCAT(OMPI_OP_TYPE_PREPEND, type##type_size####x##type_cnt##_t) vsrc, vdst;       \
+        for (; left_over >= type_cnt; left_over -= type_cnt) {                                \
+            vsrc = vld1q##_##type_name##type_size(in);                                        \
+            vdst = vld1q##_##type_name##type_size(out);                                       \
+            in += type_cnt;                                                                   \
+            vdst = OP_CONCAT(OMPI_OP_OP_PREPEND, op##q##_##type_name##type_size)(vdst, vsrc); \
+            vst1q##_##type_name##type_size(out, vdst);                                        \
+            out += type_cnt;                                                                  \
+        }                                                                                     \
+                                                                                              \
+        while (left_over > 0) {                                                               \
+            int how_much = (left_over > 8) ? 8 : left_over;                                   \
+            neon_loop(out, out, in);                                                          \
+            left_over -= how_much;                                                            \
+            out += how_much;                                                                  \
+            in += how_much;                                                                   \
+        }                                                                                     \
+    }
+#elif defined(GENERATE_SVE_CODE)
+#define OP_AARCH64_FUNC(name, type_name, type_size, type_cnt, type, op)           \
+    static void OP_CONCAT(ompi_op_aarch64_2buff_##name##_##type##type_size##_t, APPEND) \
+                            (const void *_in, void *_out, int *count,             \
+                             struct ompi_datatype_t **dtype,                      \
+                             struct ompi_op_base_module_1_0_0_t *module)          \
+    {                                                                             \
+        int types_per_step = svcnt(*((type##type_size##_t *) _in));               \
+        int left_over = *count;                                                   \
+        type##type_size##_t *in = (type##type_size##_t *) _in,                    \
+                            *out = (type##type_size##_t *) _out;                  \
+        OP_CONCAT(OMPI_OP_TYPE_PREPEND, type##type_size##_t) vsrc, vdst;          \
+        svbool_t pred = svptrue_b##type_size();                                   \
+        for (; left_over >= types_per_step; left_over -= types_per_step) {        \
```

Comment:
Why not do the predicate logic in the main loop and skip the cleanup?

```
svbool_t pred;
for (; left_over >= types_per_step; left_over -= types_per_step) {
    pred = svwhilelt_b##type_size(0, leftover);
    // loop body
}
```

Reply:
Here's some documentation on using SVE predicates for loop control: https://developer.arm.com/documentation/100891/0610/coding-considerations/using-sve-intrinsics-directly-in-your-c-code

Reply:
Because then I would have to recompute the predicate at each iteration, while here I can precompute how many iterations I need. Do you think there is any performance benefits ?

Reply:
The predicates are computed with the integer pipelines leaving the vector pipelines free to do the work you care about. Recomputing the predicate every iteration is the intended way to use SVE.

Reply:
Doing it that way will result in fewer branches and less code which lower the pressure of the front end of the core. It probably won't have a huge impact on performance, but less code is generally better.

Reply:
sounds good. let me change it.

Reply:
I looked at the updated SVE loops. What you have written looks good to me!